packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1442,14 +1445,29 @@ export class KeyringController extends BaseController<
     encryptionKey: string,
     encryptionSalt: string,
   ): Promise<void> {
-    return this.#withRollback(async () => {
-      this.#keyrings = await this.#unlockKeyrings(
+    const { newMetadata } = await this.#withRollback(async () => {
+      const result = await this.#unlockKeyrings(
         undefined,
         encryptionKey,
         encryptionSalt,
       );
       this.#setUnlocked();
+      return result;
     });
+
+    try {
+      // if new metadata has been generated during login, we
+      // can attempt to upgrade the vault.
+      await this.#withRollback(async () => {
+        if (newMetadata) {
+          await this.#updateVault();
```

Comment:
Do we have a test yet for this scenario?

Reply:
Good point, it was missing for this method! I added a test for `submitPassword` in https://github.com/MetaMask/core/pull/5725/commits/8aed3547d606c91087196e689dc9c243b94d5909 and one for `submitEncryptionKey` in [c71d8a6](https://github.com/MetaMask/core/pull/5725/commits/c71d8a6a85f5850b7e0fdb9987a2d05ab22522aa)